src/transform/src/attribute/analysis.rs

```diff
@@ -0,0 +1,441 @@
+                        bundle.analyse(expr, &result);
+                        result.analyses.insert(*id, bundle);
+                    }
+                }
+            }
+
+            result
+        }
+    }
+
+    /// An abstraction for an analysis and associated state.
+    trait AnalysisBundle: Any {
+        fn analyse(&mut self, expr: &MirRelationExpr, depends: &Derived);
+        /// Upcasts `self` to a `&dyn Any`.
+        ///
+        /// NOTE: This is required until <https://github.com/rust-lang/rfcs/issues/2765> is fixed
+        fn as_any(&self) -> &dyn std::any::Any;
+    }
+
+    /// A wrapper for analysis state.
+    struct Bundle<A: Analysis> {
+        results: Vec<A::Value>,
+    }
+
+    impl<A: Analysis> AnalysisBundle for Bundle<A> {
+        fn analyse(&mut self, expr: &MirRelationExpr, depends: &Derived) {
+            let value = A::derive(expr, &self.results[..], depends);
+            self.results.push(value);
+        }
+        fn as_any(&self) -> &dyn std::any::Any {
+            self
+        }
+    }
+}
+
+/// Expression subtree sizes
+///
+/// This analysis counts the number of expressions in each subtree, and is most useful
+/// for navigating the results of other analyses that are offset by subtree sizes.
+pub mod subtree {
+
+    use super::{Analysis, Derived};
+    use mz_expr::MirRelationExpr;
+
+    /// Analysis that determines the size in child expressions of relation expressions.
+    #[derive(Debug)]
+    pub struct SubtreeSize;
+
+    impl Analysis for SubtreeSize {
+        type Value = usize;
+
+        fn derive(
+            expr: &MirRelationExpr,
+            results: &[Self::Value],
+            _depends: &Derived,
+        ) -> Self::Value {
+            match expr {
+                MirRelationExpr::Constant { .. } | MirRelationExpr::Get { .. } => 1,
+                _ => {
+                    let n = results.len();
+                    let mut offset = 1;
+                    for _ in expr.children() {
+                        offset += results[n - offset];
+                    }
+                    offset
+                }
+            }
+        }
+    }
+}
+
+/// Expression arities
+mod arity {
+
+    use super::subtree::SubtreeSize;
+    use super::{Analysis, Derived, DerivedBuilder};
+    use mz_expr::MirRelationExpr;
+
+    /// Analysis that determines the number of columns of relation expressions.
+    #[derive(Debug)]
+    pub struct Arity;
+
+    impl Analysis for Arity {
+        type Value = usize;
+
+        fn announce_dependencies(builder: &mut DerivedBuilder) {
+            builder.require::<SubtreeSize>();
+        }
+
+        fn derive(
+            expr: &MirRelationExpr,
+            results: &[Self::Value],
+            depends: &Derived,
+        ) -> Self::Value {
+            let position = results.len();
+            let mut offsets = depends
+                .children_of_rev(position, expr.children().count())
+                .expect("SubtreeSize missing")
```

Comment:
I agree that `subtree` should be somehow baked in as a default, especially if it allows us to make the result type of `children_of_rev` non-optional and get rid of these `except` calls.

Reply:
It has been made so!